go/graphd/connect.go

```diff
@@ -0,0 +1,192 @@
+	return connCh
+}
+
+// dial attempts to acquire a connection to url with the specified timeout.  On connection
+// success or failure, dial will attempt to send the net.Conn over the result channel if it
+// can read from the awaitingConn channel (used as mutex here).  If the read fails (awaitingConn
+// has been closed by the caller), dial will attempt to close the acquired connection.
+func (g *graphd) dial(url *url.URL, timeout time.Duration, connCh connChan) {
+	// A timeout of zero value is equivalent to no timeout.
+	dialer := net.Dialer{Timeout: timeout}
+
+	// Attempt to dial.  Log error, but send conn anyway.  A nil conn will be discarded by the
+	// caller.
+	g.LogDebugf("dialing %v with timeout %v", url, timeout)
+	conn, err := dialer.Dial(url.Scheme, url.Host)
+	if err != nil {
+		g.LogErrf("failed to dial %v: %v", url, err)
+	}
+
+	// If an acquired connection has not yet been sent by a dilaer, go ahead and send the conn,
+	// regardless if it's nil, and return.
+	if _, ok := <-connCh.awaitingConn; ok {
+		if conn != nil {
+			g.LogDebugf("sending acquired connection to %v on channel", url)
+		} else {
+			g.LogDebugf("sending nil connection to %v on channel", url)
+		}
+		connCh.res <- conn
+		return
+	}
+
+	// A connection was already acquired by another dialer.  If we did acquire a connection,
+	// close it.
+	if conn != nil {
+		g.LogDebugf("discarding and closing acquired connection to %v", url)
+		if err := conn.Close(); err != nil {
+			g.LogErrf("failed to close acquired connection to %v, resource leak", url)
+		}
+	}
+}
+
+// isConnected checks if this graphd instance is connected to a graphd database.
+// TODO: Currently only a stub.  True case needs a write check (ie. "status()").
+func (g *graphd) isConnected() bool {
+	if g.conn == nil {
+		return false
+	}
+
+	return true
+}
+
+// Connect to a graphd database.  Connect will attempt to connect to all URLs found in the URLs
+// list associated with this graphd instance, retaining the first successful connection.  On failure,
+// an appropriate error code is returned.  If an acquired connection is already present and valid,
+// Connect returns nil.  Timeout is specified in seconds.  A timeout of 0 is treated as no timeout.
+func (g *graphd) Connect(t int) error {
+	// Set timeout if t > 0, otherwise use the zero value (0s) which signals no timeout.
+	var timeout time.Duration
+	if t > 0 {
+		timeout = time.Duration(t) * time.Second
+	}
+
+	// If URLs list is empty, return error.
+	if len(g.urls) == 0 {
+		errStr := fmt.Sprintf("no URL found in %v", g.urls)
+		g.LogErrf(errStr)
+		return errors.New(errStr)
+	}
+
+	// If already connected, return success.
+	if g.isConnected() {
+		g.LogDebugf("already connected to %v", g.conn.RemoteAddr())
+		return nil
+	}
+
+	// In parallel, for each URL in the list, send a connection request along with the timeout.
+	g.LogDebugf("attempting to connect to %v", g.urls)
+	connCh := initConnCh()
+	numDialers := 0
```

Comment:
this looks sort of like a [sync.WaitGroup](https://golang.org/pkg/sync/#WaitGroup) - you're launching a go routine and you want to wait for a thing to happen? Though i guess down below you're actually assigning the connection...

don't you want to capture all connections, though? Why is line 134 not `g.conn = append(g.conn, conn)` ?